community/bolt/src/test/java/org/neo4j/bolt/transport/TransportWriteThrottleTest.java

```diff
@@ -0,0 +1,212 @@
+        when( context.channel() ).thenReturn( channel );
+    }
+
+    @Test
+    public void shouldSetWriteBufferWatermarkOnChannelConfigWhenInstalled()
+    {
+        // given
+        TransportThrottle throttle = newThrottle();
+
+        // when
+        throttle.install( channel );
+
+        // expect
+        ArgumentCaptor<WriteBufferWaterMark> argument = ArgumentCaptor.forClass( WriteBufferWaterMark.class );
+        verify( config, times( 1 ) ).setWriteBufferWaterMark( argument.capture() );
+
+        assertEquals( 64, argument.getValue().low() );
+        assertEquals( 256, argument.getValue().high() );
+    }
+
+    @Test
+    public void shouldNotLockWhenWritable() throws Exception
+    {
+        // given
+        TransportThrottle throttle = newThrottleAndInstall( channel );
+        when( channel.isWritable() ).thenReturn( true );
+
+        // when
+        Future future = Executors.newSingleThreadExecutor().submit( () -> throttle.acquire( channel ) );
+
+        // expect
+        try
+        {
+            future.get( 2000, TimeUnit.MILLISECONDS );
+        }
+        catch ( Throwable t )
+        {
+            fail( "should not throw" );
+        }
+
+        assertTrue( future.isDone() );
+        verify( lock, never() ).lock( any(), anyLong() );
+    }
+
+    @Test
+    public void shouldLockWhenNotWritable() throws Exception
+    {
+        // given
+        TransportThrottle throttle = newThrottleAndInstall( channel );
+        when( channel.isWritable() ).thenReturn( false );
+
+        // when
+        Future future = Executors.newSingleThreadExecutor().submit( () -> throttle.acquire( channel ) );
+
+        // expect
+        try
+        {
+            future.get( 2000, TimeUnit.MILLISECONDS );
+
+            fail( "should timeout" );
+        }
+        catch ( TimeoutException t )
+        {
+            // expected
+        }
+        catch ( Throwable t )
+        {
+            fail( "should timeout" );
+        }
+
+        assertFalse( future.isDone() );
+        verify( lock, atLeastOnce() ).lock( any(), anyLong() );
+        verify( lock, never() ).unlock( any() );
+    }
+
+    @Test
+    public void shouldResumeWhenWritableOnceAgain() throws Exception
+    {
+        // given
+        TransportThrottle throttle = newThrottleAndInstall( channel );
+        when( channel.isWritable() ).thenReturn( false ).thenReturn( true );
+
+        // when
+        throttle.acquire( channel );
+
+        // expect
+        verify( lock, atLeastOnce() ).lock( any(), anyLong() );
+        verify( lock, never() ).unlock( any() );
+    }
+
+    @Test
+    public void shouldResumeWhenWritabilityChanged() throws Exception
+    {
+        // given
+        TransportThrottle throttle = newThrottleAndInstall( channel );
+        when( channel.isWritable() ).thenReturn( false );
+
+        Future future = Executors.newSingleThreadExecutor().submit( () -> throttle.acquire( channel ) );
+        Thread.sleep( 500 );
```

Comment:
Sleep might make this test flaky. Would it be ok to do `future.get()` here instead?

Reply:
That future will not run to completion until the second part of the test runs, it will be blocked because of ```channel#isWritable``` returning ```false```. That was the reason I chose ```Thread#sleep```. Any other option you would recommend?

Reply:
Hm, maybe expose number of waiters in test impl of `ThrottleLock` somehow and wait for this number to turn 1 in this test?

Reply:
Oh, no, sorry. You're right, it doesn't block in the test, ```lock``` is merely a mock.